### DISCUSSION TITLE
Streamlink API read method looping

### DISCUSSION THREAD

**Discussion by aswinme** [streamlink/streamlink#4690] [Q&A] [answered]
Hello coders!..I am trying to use streamlink api for downloading live youtube videos. I am stuck with the looping of read method to write to mp4 file. Please see my code below. 
```
import streamlink
from streamlink import Streamlink
from functools import partial

session = Streamlink()
streams = session.streams("https://www.youtube.com/watch?v=-slOkFqaWlM")
with open("video3.mp4","wb") as f:
    for block in iter(partial(streams.read, 100), b""): #100 is chunk size
        f.write(streams["480p"].open().read(-1))
f.close()
```
Could you please help me in figuring out what i am doing wrong here ? . I have spend days in going through streamlink api guides and discussion forums but not still able to figure out how to loop correctly.

**Comment by bastimeyer** [comment, marked as answer]
> I have spend days in going through streamlink api guides and discussion forums but not still able to figure out how to loop correctly.

Going through the actual code makes much more sense instead of reading outdated/incomplete documentation or random comments in online forums.

The code snippet you've posted doesn't make any sense and looks like random gibberish.

This is the most basic implementation for reading streams and writing the content to a file. This doesn't have any error handling or stream halting logic, it doesn't log anything, and it runs in a single thread.

```py
import sys
from functools import partial
from pathlib import Path

from streamlink import Streamlink


def run(url, quality, file):
    session = Streamlink()
    streams = session.streams(url)
    if not streams:
        return
    stream = streams.get(quality, "best")
    try:
        fd_stream = stream.open()
        with Path(file).open("wb") as fd_file:
            for chunk in iter(partial(fd_stream.read, 8192), b""):
                fd_file.write(chunk)
    finally:
        if fd_stream:
            fd_stream.close()

if __name__ == "__main__":
    run(*sys.argv[1:])
```
